app/validators/digital_object/enabled_dynamic_fields_validations.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+module DigitalObject::EnabledDynamicFieldsValidations
+  # Returns any errors related to project level validations.
+  #
+  # @return [String] if there are errors
+  # @return false if there are no errors
+  def enabled_field_errors(digital_object, data)
+    errors = []
+    dynamic_field_paths = []
+    data.each do |df_group, children|
+      dynamic_field_paths = collect_field_paths(df_group, children, dynamic_field_paths)
+      dynamic_field_paths.each do |dynamic_field_path|
+        if (e = not_enabled?(dynamic_field_path, digital_object))
+          errors.concat e.map { |i| [dynamic_field_path, i] }
+        end
+      end
+    end
+
+    errors = check_required_fields(dynamic_field_paths, digital_object, errors)
+
+    errors
+  end
+
+  def not_enabled?(dynamic_field_path, digital_object)
+    no_match = true
+    dynamic_field = DynamicField.find_by(path: dynamic_field_path)
```

Comment:
This SQL query and the one below it appear to run separately for every field present in the record, which can get expensive for digital objects with a lot of fields.  Instead, I'd recommend collecting a single list of all fields present and comparing it to a single list of all enabled fields, and seeing what doesn't match up.

Reply:
After adding that other comment with the suggested new implementation for `check_required_fields`, I'm realizing that there's a parallel here for checking enabled fields.  You may want to do something similar.  In fact, you could even retrieve all of the enabled DynamicField instances just once in the `enabled_field_errors` method, and then pass them into method and `check_required_fields`.  That would reduce the SQL queries even more.

Step 1: Get list of all EnabledDynamicFields for the project/digital_object_type combo.  Make sure to have `.includes(:dynamic_field)` on that ActiveRecord query to grab all of the Fields in one go with that query.
Step 2: Take that list of EnabledDynamicFields (which includes DynamicFields) and run `enabled_dynamic_fields.map {|edf| edf.dynamic_field } to get the list of DynamicFields, and check it against the fields present in data (i.e. replacement for not_enabled? check).
Step 3: Also take that list of EnabledDynamicFields (which includes DynamicFields) and run `enabled_dynamic_fields.select {|edf| edf.required }.map{|edf| edf.dynamic_field }` to get the list of required dynamic fields, and check it against the fields present in data (i.e. replacement for `check_required_fields`).